Update anisotropic viscosity cookbook with new parameters (#6919)
geodynamics/aspect: cookbooks/CPO_induced_anisotropic_viscosity/doc/CPO_induced_anisotropic_viscosity.md

Reviewer: It might be worth adding some documentation describing why you choose the values that you do in this example. There are a lot of parameters in this section and to me it is not clear how you arrive at a certain value (`Mobility` for example, or `Threshold GBS`). Alternatively, you could link to something that can help users understand where these values come from if it would be too difficult to briefly summarize the parameter choices in this cookbook

# CPO induced anisotropic viscosity

*This section was contributed by Yijun Wang, Ágnes Király and Theo Häußler*

This cookbook explains how to use the CPO-induced anisotropic viscosity material model to set up a shear box model.

## Introduction

Individual crystals of the mineral olivine reorganize their orientations into crystal-preferred orientations (CPO) under deformation. The viscous properties of olivine crystals are direction-dependent (anisotropic), which suggests that the effective viscosity for olivine rocks/aggregates is different when deformations occur in different directions relative to the CPO. This cookbook model computes an anisotropic viscosity based on the CPO evolution predicted by D-Rex ({cite}`fraters_billen_2021_cpo`; {cite}`kaminski2004`) and includes this information in the subsequent modeling process based on the formulation of Király at al., {cite}`kiraly:etal:2026`.

### Forward Rheology

Our constitutive equation for the relationship between the strain rate and stress using the anisotropic viscosity tensor is adapted from {cite:t}`signorelli:etal:2021`. In Kelvin notation we can write:

```{math}
:label: eqn:anisotropic_general_stress

> \dot{\varepsilon}_{i} = \gamma J(\sigma_{pq})^{(n-1)/2} A_{ij} \sigma_{j} \text{ ,}

Reviewer 2: what are the indices p and q and how do they fit into the rest of the equation? If they are different from i and j shouldnt the result of the equation have three indices p, q, and i (j is eliminated by the multiplication). Or is this already the case? I.e is epsilon a rank-2 tensor quantity and _i the remaining index?

Reviewer 3: yes, i changed the indices! Both p and q are not summed over (they also don't appear twice) and are just there to tell people that the stress \sigma is a second rank tensor. 
The equation there is written in kelvin notation. A_{ij} is a symmetric* dim 3 rank-4 tensor compressed to a dim 6 rank 2 tensor.Similarly epsilon and sigma are symmetric tensor, which are written as dim 6 column vectors.  

We could also use the notation 
```math
\dot{\varepsilon}_{ij} = \gamma J(\sigma_{pq})^{(n-1)/2} A_{ijkl} \sigma_{kl} \text{ ,}
```
which is equivalent to what is currently written. The inconsistency with this is that previously $A$ was defined in Kelvin notation ($A_{ij} =$ not $A_{ijkl} =$)

$^*$ I can write down in equations what symmetric means in this context if that's necessary.

Reviewer 2: Thanks for the explanation, but I am still not quite understanding then what p and q are represented by on the left hand side of the equation. It is not quite clear if epsilon (on the LHS) is also a tensor quantity (it probably is, but it would have the subscripts i and j, not p and q), or if J(sigma_pq) is a function that returns a scalar value, and this is how the indices disappear. In other words the indices on the LHS and right hand side seem to be not balanced.

Reviewer 3: Ahh okay thank you for spelling out your confusion with the equation.
\dot{\epsilon}_i on the left side is a vector representation of the symmetric tensor \dot{\epsilon}_{ij}. 
Physically they are the same quantity. J(\sigma_ij) returns a scalar and is based on the components of the stress \sigma_ij.
We will add a short description of kelvin notation, where we specify what we mean when writing epsilon with one index or two and remove the dependence on the stress in the equation to make it more clear.

```

where $\gamma$ is the part of fluidity (the inverse of viscosity) which is temperature- and grain-size dependent:

```{math}
:label: eqn:fluidity
\gamma=\gamma_0 exp \left(\frac{-Q}{RT} \right) /d^m \text{ .}
```

> $\gamma_0=1.1\times 10^{5}$ is the isotropic fluidity, $Q=530$ $kJ/mol$ is the activation energy, $R=8.314 m^3 \cdot Pa \cdot K^{−1} \cdot$ $mol^{−1}$ is the gas constant, $d=0.001$ $m$ is the grain size, and $m=0.73$ is the grain size exponent. These values for olivine are taken from rock experiments performed by {cite:t}`hansen:etal:2016` and {cite:t}`HK04`. $J(\sigma_{ij})$ is the yield potential, where $\sigma_{ij}$ is the (anisotropic) stress computed using the tensorial and scalar component of the anisotropic viscosity:

Reviewer: ```suggestion
$\gamma_0=1.1\times 10^{5}$ is the isotropic fluidity, $Q=530$ $kJ/mol$ is the activation energy, $R=8.314$ m$^3 \cdot$ Pa $\cdot$ K$^{−1} \ cdot$ mol$^{−1}$ is the gas constant, $d=0.001$ m is the grain size, and $m=0.73$ is the grain size exponent. These values for olivine are taken from rock experiments performed by {cite:t}`hansen:etal:2016` and {cite:t}`HK04`. $J(\sigma_{ij})$ is the yield potential, where $\sigma_{ij}$ is the (anisotropic) stress computed using the tensorial and scalar component of the anisotropic viscosity:
```

Reviewer: So that the units are not formatted as italics!

```{math}
:label: eqn:hill_coefficients
F = 0.592 a_1^2 - 0.832 a_1 - 0.001 a_2 - \frac{0.000}{a_3} + 0.380 b_1^2 - 0.533 b_1 + 0.468 b_2 - \frac{0.001}{b_3} - 1.249 c_1^2 + 1.075 c_1 - 0.168 c_2 + \frac{0.003}{c_3} + 0.52 \\
G = -1.695 a_1^2 + 1.336 a_1 - 0.184 a_2 + \frac{0.000}{a_3} + 0.750 b_1^2 + 0.691 b_1 + 0.377 b_2 - \frac{0.002}{b_3} - 0.670 c_1^2 - 0.552 c_1 - 0.428 c_2 + \frac{0.003}{c_3} + 0.26 \\
H = -1.140 a_1^2 + 1.353 a_1 + 0.751 a_2 - \frac{0.002}{a_3} - 0.256 b_1^2 - 1.006 b_1 - 0.116 b_2 + \frac{0.003}{b_3} + 0.648 c_1^2 - 0.031 c_1 - 0.080 c_2 + \frac{0.006}{c_3} + 0.75 \\
L = -3.511 a_1^2 + 2.686 a_1 + 0.360 a_2 - \frac{0.001}{a_3} + 3.948 b_1^2 - 3.816 b_1 - 0.779 b_2 + \frac{0.004}{b_3} + 4.122 c_1^2 - 2.483 c_1 - 1.320 c_2 + \frac{0.002}{c_3} + 2.00 \\
M = 4.537 a_1^2 - 3.228 a_1 + 0.276 a_2 + \frac{0.007}{a_3} - 7.447 b_1^2 + 5.764 b_1 - 1.403 b_2 - \frac{0.032}{b_3} + 2.968 c_1^2 - 3.435 c_1 - 2.266 c_2 + \frac{0.122}{c_3} + 2.44 \\
N = 7.873 a_1^2 - 7.934 a_1 - 2.588 a_2 + \frac{0.030}{a_3} + 7.606 b_1^2 - 5.469 b_1 - 0.348 b_2 + \frac{0.064}{b_3} - 1.788 c_1^2 + 2.255 c_1 + 3.023 c_2 - \frac{0.103}{c_3} + 3.70
```

### CPO Reference Frame

> Since the Hill coefficients are defined in the microscopic CPO reference frame, and parameters computed in ASPECT are in the macroscopic model reference frame, several reference frame conversions are needed. We determine the mean CPO orientation from the eigenvectors associated with the largest eigenvalues of the second-order orientation tensor (or covariance matrix) for all three symmetry axes. The corresponding eigenvalues quantify the dispersion of orientations around these mean orientation {cite}`bingham:1974`.

Reviewer: ```suggestion
Since the Hill coefficients are defined in the microscopic CPO reference frame, and parameters computed in ASPECT are in the macroscopic model reference frame, several reference frame conversions are needed. We determine the mean CPO orientation from the eigenvectors associated with the largest eigenvalues of the second-order orientation tensor (or covariance matrix) for all three symmetry axes. The corresponding eigenvalues quantify the dispersion of orientations around the mean orientation {cite}`bingham:1974`.
```

\texttt{max\_eigenvector}_{a3} & \texttt{max\_eigenvector}_{b3} & \texttt{max\_eigenvector}_{c3}
\end{matrix} \right]
```

> As sometimes the 3 primary eigenvectors do not compose an othrogonal frame, therefore a singular value decomposition is used to orthogonalize the rotationn matrix $R$. We compute the rotation matrix $R$ on the particles and further convert it to Euler angles for computation and memory efficiency. These properties need to be interpolated from particles to fields to be used in the material model. As a result, the anisotropic viscosity material model requires at least one particle in each cell so that all cells can have the texture parameters (Euler angles and eigenvalues) for constructing the rotation matrix R and compute the Hill coefficients. In the material model, the interpolated Euler angles are converted to the rotation matrix again. We use the same notation R to describe the rotation matrix used in the material model in the following paragraphs.

Reviewer: ```suggestion
As sometimes the 3 primary eigenvectors do not compose an othrogonal frame, a singular value decomposition is used to orthogonalize the rotation matrix $R$. We compute the rotation matrix $R$ on the particles and further convert it to Euler angles for computation and memory efficiency. These properties need to be interpolated from particles to fields to be used in the material model. As a result, the anisotropic viscosity material model requires at least one particle in each cell so that all cells can have the texture parameters (Euler angles and eigenvalues) for constructing the rotation matrix R and computing the Hill coefficients. In the material model, the interpolated Euler angles are converted to the rotation matrix again. We use the same notation R to describe the rotation matrix used in the material model in the following paragraphs.
```

```

$R'$ and $R_K'$ is the transpose of matrix $R$ and $R_K$ respectively. We save $\frac{1}{\gamma J(R'_{ik}\sigma_{kl}R_{lj})^{(n-1)/2}}$ as the material model viscosity output, which we call the scalar viscosity. The scalar viscosity of the current step is also stored into the prescribed field, so that at the beginning of the next step, the anisotropic stress is computed with the scalar viscosity of the previous step using {math:numref}`eqn:anisotropic_stress`. The tensorial part of anisotropic viscosity, $[R_K]_{ik}A_{kl}^{-1}[R_K']_{lj}$ is called the stress-strain director and is stored in the additional outputs. Because the scalar viscosity depends on stress and the stress is determined by the scalar viscosity from the previous time step, the computation of the scalar viscosity is potentially unstable and its value can oscillate across nonlinear iterations. This oscillation can ultimately cause a numerical instability associated with the prediction of the anisotropic viscosity. Therefore, we damp the scalar viscosity computation using a fixed-point iteration, by applying only half of the change in each iteration until the result converges.

### Analytical inversion

> The inverse rheology of an orthotropic non-linear rheology can be derived from the symmetry of the strain-rate tensor {cite}`rathmann:lilien:2022`. Writing the orthotropic rheology and its inversion in the CPO reference frame we can compare the Hill coefficients to the fluidity coefficients or enhancement factors of {cite}`rathmann:lilien:2022`. The reformulated inverse rheology in terms of Hill coefficients  in Kelvin notation then becomes:

Reviewer 2: ```suggestion
The inverse rheology of an orthotropic non-linear rheology can be derived from the symmetry of the strain-rate tensor {cite}`rathmann:lilien:2022`. Writing the orthotropic rheology and its inversion in the CPO reference frame we can compare the Hill coefficients to the fluidity coefficients or enhancement factors of {cite}`rathmann:lilien:2022`. The reformulated inverse rheology in terms of Hill coefficients in Kelvin notation then becomes:
```

        -2H_2-2H_3 +H_1  & 0 & 0 & 0 \\
    & & 4H_3+H_1+H_2 & 0 & 0 & 0 \\
        &       &       & \frac{9\Gamma}{4H_4} & 0 & 0 \\
        &       &       &   & \frac{9\Gamma}{4H_5} & 0 \\
 \text{sym} & & & & & \frac{9\Gamma}{4H_6}
    \end{pmatrix} \text{ ,}
\end{aligned}
```
with the anisotropic tensor for viscosity in Kelivin notation $V_{ij}$ and the non-linear viscosity $\eta(\dot{\varepsilon}_I)$, which is based on an anisotropic strain-rate invariant $\dot{\varepsilon}_I$:
```{math}
:label: eqn:anisotropic_viscosity
\begin{aligned}
    \eta(\dot{\varepsilon}_I) &= \gamma^{-1/n}\dot{\varepsilon}_I^{(1-n)/n} \\
    \dot{\varepsilon}_I &= \left[\sum_i\frac{2}{3\Gamma}(\dot{\varepsilon}_{ii}^2(4H_{i}+H_{j_i}+H_{k_i}) + 2\dot{\varepsilon}_{j_ij_i}\dot{\varepsilon}_{k_ik_i}(H_i-2H_{j_i}-2H_{k_i}))+ 3/H_{i+3}\dot{\varepsilon}_{j_ik_i}^2\right]^{1/2} \text{ .}
\end{aligned}
```
Here we use the convention $H_1=F,H_2=G,H_3=H$, $H_4=L,H_5=M,H_6=N$ and the shifted indices $j=\{1,2,0\}$ and $k=\{2,0,1\}$. $\Gamma$ ($\gamma$ in {cite}`rathmann:lilien:2022`) is a short script for:
```{math}
:label: eqn:factor_of_coefficients
\Gamma =  4\sum_i H_{j_i}H_{k_i} \text{ .}
```

> In the isotropic case ($H_i=0.5$, $H_{i+3}=1.5$) we then have a flowlaw of the form,

Reviewer 2: ```suggestion
In the isotropic case ($H_i=0.5$, $H_{i+3}=1.5$) we then have a flow law of the form,
```

```{literalinclude} min_particles_per_cell.part.prm
```

> - **CPO particle property**: The CPO particle property must be stored for use by the AV model. This requires enabling the particle and crystal preferred orientation postprocessors and the relevant subsections for them, including the CPO Bingham Average plugin, which calculates the Hill coefficients. We set the mobility, nucleation efficiently, and threshold GBS values following {cite:t}`hansen:etal:2016`, for grain boundary mobility, grain nucleation, and grain boundary sliding, respectively.

Reviewer 2: ```suggestion
- **CPO particle property**: The CPO particle property must be stored for use by the AV model. This requires enabling the particle and crystal preferred orientation postprocessors and the relevant subsections for them, including the CPO Bingham Average plugin, which calculates the Hill coefficients. We set the mobility, nucleation efficiency, and threshold GBS values following {cite:t}`hansen:etal:2016`, for grain boundary mobility, grain nucleation, and grain boundary sliding, respectively.
```

```{literalinclude} compositional_field.part.prm
```

> In the `CPO induced Anisotropic Viscosity` material model subsection, all parameters have reasonable default values and do not need to be manually specified unless customization is needed. For non-dimensional cases it is possible to define the Grain size $d$, the Grain size exponent $m$, the Fluidity constant $\gamma_0$, the Activation energy $Q$, the non-linear stress exponent $n$ and if the analytical inversion or the iterative inversion should be used.

Reviewer 2: You probably meant the actual input parameters, but then put quotation marks around the parameter names (or remove the capitalization like I did here).

```suggestion
In the `CPO induced Anisotropic Viscosity` material model subsection, all parameters have reasonable default values and do not need to be manually specified unless customization is needed. For non-dimensional cases it is possible to define the grain size $d$, the grain size exponent $m$, the fluidity constant $\gamma_0$, the activation energy $Q$, the non-linear stress exponent $n$ and if the analytical inversion or the iterative inversion should be used.
```
